File: project_euler/Problem074/Problem074.py
import math
from typing import List

class Problem074:
    def solution(self) -> None:
        target, border = 60, 1000000
        step, permutations = int(math.log10(border)), []
        for i in range(1, 10):
            permutations.append(i)
        chains, terms, factorials = set(), {}, [0 for _ in range(10)]
        factorials[0] = 1
        for i in range(1, 10):
            factorials[i] = factorials[i - 1] * i
        counts, auxiliary, result = [], [0 for _ in range(10)], 0
        for value in permutations:
            if target == self.__get_terms(value, chains, terms, factorials, counts):
                result += self.__get_permutations(value, factorials, counts)
        for i in range(1, step):
            self.__get_digits(permutations, auxiliary)
            for value in permutations:
                if target == self.__get_terms(value, chains, terms, factorials, counts):
                    result += self.__get_permutations(value, factorials, counts)
        print(result)

    def __get_terms(self, value: int, chains: set[int], terms: dict[int, int], factorials: List[int], counts: List[int]) -> int:
        chains.clear()
        pivot = value
        while True:
            if value in terms:
                result = len(chains) + terms[value]
                break
            if value in chains:
                result = len(chains)
                break

            chains.add(value)
            value = self.__get_summation(value, factorials, counts)
        terms[pivot] = result
        return result

    def __get_permutations(self, value: int, factorials: List[int], counts: List[int]) -> int:
        length, exists_0 = 0, False
        counts[:] = [0 for _ in range(10)]
        while value:
            length += 1
            remainder = value % 10
            if not remainder:
                exists_0 = True
            counts[remainder] += 1
            value //= 10

        result = factorials[length]
        if exists_0:
            result = result * (length - 1) // length
        for count in counts:
            if 1 != count:
                result //= factorials[count]
        return result

    def __get_digits(self, permutations: List[int], auxiliary: List[int]) -> None:
        auxiliary.extend(permutations)
        permutations.clear()
        for value in auxiliary:
            remainder = value % 10
            for i in range(remainder + 1):
                permutations.append(value * 10 + i)
        auxiliary.clear()

    def __get_summation(self, value: int, factorials: List[int], counts: List[int]) -> int:
        result = 0
        while value:
            result += factorials[value % 10]
            value //= 10

        counts[:] = [0 for _ in range(10)]
        while result:
            counts[9 - result % 10] += 1
            result //= 10

        for i, count in enumerate(counts):
            for j in range(count):
                result = result * 10 + 9 - i
        return result

File: project_euler/Problem074/Problem074.java
package project_euler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Problem074 {
    public void solution() {
        int target = 60;
        int border = 1000000;
        int step = (int) Math.log10(border);
        List<Integer> permutations = new ArrayList<>();
        for (int i = 1; i < 10; ++i) {
            permutations.add(i);
        }
        Set<Integer> chains = new HashSet<>();
        Map<Integer, Integer> terms = new HashMap<>();
        int[] factorials = new int[10];
        factorials[0] = 1;
        for (int i = 1; i < 10; ++i) {
            factorials[i] = factorials[i - 1] * i;
        }
        int[] counts = new int[10];
        List<Integer> auxiliary = new ArrayList<>();
        int result = 0;
        for (int value : permutations) {
            if (target == getTerms(value, chains, terms, factorials, counts)) {
                result += getPermutations(value, factorials, counts);
            }
        }
        for (int i = 1; i < step; ++i) {
            getDigits(permutations, auxiliary);
            for (int value : permutations) {
                if (target == getTerms(value, chains, terms, factorials, counts)) {
                    result += getPermutations(value, factorials, counts);
                }
            }
        }
        System.out.println(result);
    }

    private int getTerms(int value, Set<Integer> chains, Map<Integer, Integer> terms, int[] factorials, int[] counts) {
        chains.clear();
        int pivot = value;
        int result;
        while (true) {
            if (terms.containsKey(value)) {
                result = chains.size() + terms.get(value);
                break;
            }
            if (chains.contains(value)) {
                result = chains.size();
                break;
            }

            chains.add(value);
            value = getSummation(value, factorials, counts);
        }
        terms.put(pivot, result);
        return result;
    }

    private int getPermutations(int value, int[] factorials, int[] counts) {
        int length = 0;
        boolean exists0 = false;
        Arrays.fill(counts, 0);
        while (0 != value) {
            ++length;
            int remainder = value % 10;
            if (0 == remainder) {
                exists0 = true;
            }
            ++counts[remainder];
            value /= 10;
        }

        int result = factorials[length];
        if (exists0) {
            result = result * (length - 1) / length;
        }
        for (int count : counts) {
            if (1 != count) {
                result /= factorials[count];
            }
        }
        return result;
    }

    private void getDigits(List<Integer> permutations, List<Integer> auxiliary) {
        auxiliary.addAll(permutations);
        permutations.clear();
        for (int value : auxiliary) {
            int remainder = value % 10;
            for (int i = 0; i <= remainder; ++i) {
                permutations.add(value * 10 + i);
            }
        }
        auxiliary.clear();
    }

    private int getSummation(int value, int[] factorials, int[] counts) {
        int result = 0;
        while (0 != value) {
            result += factorials[value % 10];
            value /= 10;
        }

        Arrays.fill(counts, 0);
        while (0 != result) {
            ++counts[9 - result % 10];
            result /= 10;
        }

        for (int i = 0; i < 10; ++i) {
            int count = counts[i];
            for (int j = 0; j < count; ++j) {
                result = result * 10 + 9 - i;
            }
        }
        return result;
    }
}
